File: backend/app/routers/classes.py
#app/routers/classes.py
from uuid import UUID
import asyncio, logging
from fastapi import APIRouter, Depends, status, HTTPException
from sqlalchemy.orm import Session

from app.db import get_db
from app.core.security import get_current_user
from app.models.user import User
from app.models.uml import Clase, Diagram
from app.schemas.clase import ClaseCreate, ClaseUpdate
from app.schemas.clase_completa import ClaseCompletaOut
from ._helpers import get_my_diagram, get_my_class
from app.utils import realtime_events  # 👈 notificaciones en tiempo real

logger = logging.getLogger(__name__)

router = APIRouter(prefix="/diagrams", tags=["classes"])


def _ensure_unique_class_name(db: Session, diagram_id: UUID, name: str, exclude_id: UUID | None = None):
    """Valida que no exista otra clase con el mismo nombre en el diagrama."""
    q = db.query(Clase).filter(
        Clase.diagram_id == diagram_id,
        Clase.nombre == name,
    )
    if exclude_id:
        q = q.filter(Clase.id != exclude_id)
    if q.first():
        logger.warning(f"⚠️ Clase duplicada -> diagram_id={diagram_id}, name='{name}'")
        raise HTTPException(
            status_code=400,
            detail=f"La clase '{name}' ya existe en este diagrama"
        )


# 🔹 Crear clase
@router.post("/{diagram_id}/classes", response_model=ClaseCompletaOut, status_code=status.HTTP_201_CREATED)
async def create_class(
    diagram_id: UUID,
    body: ClaseCreate,
    db: Session = Depends(get_db),
    me: User = Depends(get_current_user),
):
    logger.info(f"➕ [CREATE] clase -> diagram_id={diagram_id}, user={me.id}, body={body}")
    d = get_my_diagram(db, me, diagram_id)

    _ensure_unique_class_name(db, d.id, body.name)

    try:
        c = Clase(
            nombre=body.name,
            diagram_id=d.id,
            x_grid=body.x_grid or 0,
            y_grid=body.y_grid or 0,
            w_grid=body.w_grid or 12,
            h_grid=body.h_grid or 6,
            z_index=body.z_index or 0,
        )
        db.add(c); db.commit(); db.refresh(c)
        logger.info(f"✅ Clase creada -> class_id={c.id}, diagram_id={d.id}")

        asyncio.create_task(realtime_events.notify_class_created(d.id, c))
        return c
    except Exception as e:
        logger.error(f"❌ Error creando clase -> diagram_id={diagram_id}, user={me.id}, error={str(e)}")
        raise


# 🔹 Obtener clase por ID
@router.get("/classes/{class_id}", response_model=ClaseCompletaOut)
def get_class(
    class_id: UUID,
    db: Session = Depends(get_db),
    me: User = Depends(get_current_user),
):
    logger.info(f"📥 [GET] clase -> class_id={class_id}, user={me.id}")
    c = get_my_class(db, me, class_id)
    if not c:
        logger.warning(f"⚠️ Clase no encontrada -> class_id={class_id}, user={me.id}")
    return c


# 🔹 Listar clases de un diagrama
@router.get("/{diagram_id}/classes", response_model=list[ClaseCompletaOut])
def list_classes(
    diagram_id: UUID,
    db: Session = Depends(get_db),
    me: User = Depends(get_current_user),
):
    logger.info(f"📥 [LIST] clases -> diagram_id={diagram_id}, user={me.id}")
    d = get_my_diagram(db, me, diagram_id)
    items = db.query(Clase).filter(Clase.diagram_id == d.id).all()
    logger.info(f"✅ {len(items)} clases encontradas en diagram_id={diagram_id}")
    return items


# 🔹 Actualizar clase
@router.patch("/classes/{class_id}", response_model=ClaseCompletaOut)
async def update_class(
    class_id: UUID,
    body: ClaseUpdate,
    db: Session = Depends(get_db),
    me: User = Depends(get_current_user),
):
    logger.info(f"✏️ [UPDATE] clase -> class_id={class_id}, user={me.id}, body={body}")
    c = get_my_class(db, me, class_id)

    try:
        if body.name is not None and body.name != c.nombre:
            _ensure_unique_class_name(db, c.diagram_id, body.name, exclude_id=c.id)
            c.nombre = body.name

        if body.x_grid is not None:  c.x_grid  = body.x_grid
        if body.y_grid is not None:  c.y_grid  = body.y_grid
        if body.w_grid is not None:  c.w_grid  = body.w_grid
        if body.h_grid is not None:  c.h_grid  = body.h_grid
        if body.z_index is not None: c.z_index = body.z_index

        db.commit(); db.refresh(c)
        logger.info(f"✅ Clase actualizada -> class_id={c.id}, diagram_id={c.diagram_id}")

        asyncio.create_task(realtime_events.notify_class_updated(c.diagram_id, c))
        return c
    except Exception as e:
        logger.error(f"❌ Error actualizando clase -> class_id={class_id}, error={str(e)}")
        raise


# 🔹 Eliminar clase
@router.delete("/classes/{class_id}", status_code=204)
async def delete_class(
    class_id: UUID,
    db: Session = Depends(get_db),
    me: User = Depends(get_current_user),
):
    logger.info(f"🗑️ [DELETE] clase -> class_id={class_id}, user={me.id}")
    c = get_my_class(db, me, class_id)
    diagram_id = c.diagram_id
    class_id_val = c.id

    try:
        db.delete(c); db.commit()
        logger.info(f"✅ Clase eliminada -> class_id={class_id_val}, diagram_id={diagram_id}")

        asyncio.create_task(realtime_events.notify_class_deleted(diagram_id, class_id_val))
        return
    except Exception as e:
        logger.error(f"❌ Error eliminando clase -> class_id={class_id}, error={str(e)}")
        raise

File: frontend/src/api/classes.js
import api from "./client";

// ===== Helper para mapear nombre → name =====
const mapClass = (c) => ({
  id: c.id,
  name: c.nombre,
  x_grid: c.x_grid,
  y_grid: c.y_grid,
  w_grid: c.w_grid,
  h_grid: c.h_grid,
  z_index: c.z_index,
});

// ====== CLASES ======

export const listClasses = async (diagramId) => {
  console.log("📥 [listClasses] GET", diagramId);
  try {
    const r = await api.get(`/diagrams/${diagramId}/classes`);
    console.log("✅ [listClasses] response:", r.data);
    return (r.data || []).map(mapClass);
  } catch (err) {
    console.error("❌ [listClasses] error:", err?.response?.data || err);
    throw err;
  }
};

export const createClass = async (
  diagramId,
  { name, x_grid, y_grid, w_grid, h_grid, z_index } = {}
) => {
  const body = { name, x_grid, y_grid, w_grid, h_grid, z_index };
  console.log("📤 [createClass] POST body:", body);
  try {
    const r = await api.post(`/diagrams/${diagramId}/classes`, body);
    console.log("✅ [createClass] response:", r.data);
    return mapClass(r.data);
  } catch (err) {
    console.error("❌ [createClass] error:", err?.response?.data || err);
    throw err;
  }
};

export const updateClass = async (classId, patch) => {
  console.log("📤 [updateClass] PATCH body:", patch);
  try {
    const r = await api.patch(`/diagrams/classes/${classId}`, patch);
    console.log("✅ [updateClass] response:", r.data);
    return mapClass(r.data);
  } catch (err) {
    console.error("❌ [updateClass] error:", err?.response?.data || err);
    throw err;
  }
};

export const deleteClass = async (classId) => {
  console.log("🗑️ [deleteClass] DELETE", classId);
  try {
    const r = await api.delete(`/diagrams/classes/${classId}`);
    console.log("✅ [deleteClass] OK");
    return r.data;
  } catch (err) {
    console.error("❌ [deleteClass] error:", err?.response?.data || err);
    throw err;
  }
};

// ====== HELPERS DE CLASE ======

export const updateClassPosition = (classId, { x_grid, y_grid }) =>
  updateClass(classId, { x_grid, y_grid });

export const updateClassSize = (classId, { w_grid, h_grid, x_grid, y_grid }) =>
  updateClass(classId, { w_grid, h_grid, ...(x_grid !== undefined ? { x_grid } : {}), ...(y_grid !== undefined ? { y_grid } : {}) });

export const updateClassZ = (classId, z_index) =>
  updateClass(classId, { z_index });

// ====== ATRIBUTOS ======

export const listAttributes = async (classId) => {
  console.log("📥 [listAttributes] GET", classId);
  try {
    const r = await api.get(`/diagrams/classes/${classId}/attributes`);
    console.log("✅ [listAttributes] response:", r.data);
    return r.data;
  } catch (err) {
    console.error("❌ [listAttributes] error:", err?.response?.data || err);
    throw err;
  }
};

export const createAttribute = async (classId, { name, type, required = false }) => {
  const body = { name, type, required };
  console.log("📤 [createAttribute] POST body:", body);
  try {
    const r = await api.post(`/diagrams/classes/${classId}/attributes`, body);
    console.log("✅ [createAttribute] response:", r.data);
    return r.data;
  } catch (err) {
    console.error("❌ [createAttribute] error:", err?.response?.data || err);
    throw err;
  }
};

export const updateAttribute = async (attrId, patch) => {
  console.log("📤 [updateAttribute] PATCH body:", patch);
  try {
    const r = await api.patch(`/diagrams/attributes/${attrId}`, patch);
    console.log("✅ [updateAttribute] response:", r.data);
    return r.data;
  } catch (err) {
    console.error("❌ [updateAttribute] error:", err?.response?.data || err);
    throw err;
  }
};

export const deleteAttribute = async (attrId) => {
  console.log("🗑️ [deleteAttribute] DELETE", attrId);
  try {
    const r = await api.delete(`/diagrams/attributes/${attrId}`);
    console.log("✅ [deleteAttribute] OK");
    return r.data;
  } catch (err) {
    console.error("❌ [deleteAttribute] error:", err?.response?.data || err);
    throw err;
  }
};

// ====== MÉTODOS ======

export const listMethods = async (classId) => {
  console.log("📥 [listMethods] GET", classId);
  try {
    const r = await api.get(`/diagrams/classes/${classId}/methods`);
    console.log("✅ [listMethods] response:", r.data);
    return r.data;
  } catch (err) {
    console.error("❌ [listMethods] error:", err?.response?.data || err);
    throw err;
  }
};

export const createMethod = async (classId, { name, return_type = "void" }) => {
  const body = { name, return_type };
  console.log("📤 [createMethod] POST body:", body);
  try {
    const r = await api.post(`/diagrams/classes/${classId}/methods`, body);
    console.log("✅ [createMethod] response:", r.data);
    return r.data;
  } catch (err) {
    console.error("❌ [createMethod] error:", err?.response?.data || err);
    throw err;
  }
};

export const updateMethod = async (methodId, patch) => {
  console.log("📤 [updateMethod] PATCH body:", patch);
  try {
    const r = await api.patch(`/diagrams/methods/${methodId}`, patch);
    console.log("✅ [updateMethod] response:", r.data);
    return r.data;
  } catch (err) {
    console.error("❌ [updateMethod] error:", err?.response?.data || err);
    throw err;
  }
};

export const deleteMethod = async (methodId) => {
  console.log("🗑️ [deleteMethod] DELETE", methodId);
  try {
    const r = await api.delete(`/diagrams/methods/${methodId}`);
    console.log("✅ [deleteMethod] OK");
    return r.data;
  } catch (err) {
    console.error("❌ [deleteMethod] error:", err?.response?.data || err);
    throw err;
  }
};
